name-service/program/src/processor.rs

```diff
@@ -124,10 +124,23 @@ impl Processor {
 
         let name_account = next_account_info(accounts_iter)?;
         let name_update_signer = next_account_info(accounts_iter)?;
+        let parent_name = next_account_info(accounts_iter).ok();
 
         let name_record_header = NameRecordHeader::unpack_from_slice(&name_account.data.borrow())?;
 
+        let mut is_parent_owner = false;
+
         // Verifications
+        if let Some(parent_name_acc) = parent_name {
+            if name_record_header.parent_name != *parent_name_acc.key {
+                msg!("Invalid parent name account");
+                return Err(ProgramError::InvalidArgument);
+            }
+
+            let parent_name_record_header =
+                NameRecordHeader::unpack_from_slice(&parent_name_acc.data.borrow())?;
+            is_parent_owner = parent_name_record_header.owner == *name_update_signer.key
+        }
```

Comment:
I'd prefer to avoid a `mut` here if it's all the same to you:
```suggestion
        let is_parent_owner = if let Some(parent_name_acc) = parent_name {
            if name_record_header.parent_name != *parent_name_acc.key {
                msg!("Invalid parent name account");
                return Err(ProgramError::InvalidArgument);
            }

            let parent_name_record_header =
                NameRecordHeader::unpack_from_slice(&parent_name_acc.data.borrow())?;
            parent_name_record_header.owner == *name_update_signer.key
        } else { 
            false
        };
```

Reply:
Also I prefer the style of reusing the variable name when stripping `Some` for quicker code scanning:
```
if let Some(parent_name) = parent_name { ...
```

Reply:
I agree it's better this way!